### ISSUE TITLE
Create a tiered storage implementation for TWCS

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylladb#9509]
*Installation details*
Scylla version: c4328ffc4dce9e7463be7ff54e89463a2545f380

*Description*
Many times people are asking for a tired storage feature where data would be stored at locations backed up by a different types of storage according to a certain (data) criteria.

One of the most common use cases is a time series data that never expires and users want to keep the most recent data on a fast storage (e.g. NVMe) and the older data on a slow storage (EBS or even S3 in AWS terms).

While it may be challenging to implement this for every compaction strategy it should be quite straight forward with a TWCS.

Here is a proposal for a design:

1) Implement a support for multiple `data` directories (C* already supports this today).

In the scylla.yaml it would look like something like this:
```
data_file_directories:
    - /var/lib/scylla/fast_data
    - /var/lib/scylla/slow_data
```

2) Add a configuration for a TWCS that would make it store data not older than a specific time point to the first directory in the list and push the older one to second, third and so on entries in the directories list. 

The compaction config value could be: 

```
compaction = {
  'class' : 'TimeWindowCompactionStrategy',
  'compaction_tier_threshold_unit' : string, <--- e.g. 'DAY', 'MONTH', 'MINUTE'
  'compaction_tier_threshold_size' : int, 
```

On a user side one would mount the NVMe backed mount as a first entry in the `data` list and a slower storage backed ones as the second and lower entries in that list.

And TWCS would store SSTable with the data beyond the configured threshold in the second and lower `data` dirs while keeping the new ones in the topmost one.

**Comment by dorlaor** [comment]
@raphaelsc what's your take on this? Glauber had another, a bit more generalized and complex idea

**Comment by raphaelsc** [comment]
> @raphaelsc what's your take on this? Glauber had another, a bit more generalized and complex idea

@dorlaor there are two important factors when deciding to move data from fast to slow storage:

1) Write amplification
2) Data hotness

Write amplification depends on compaction strategy. Data hotness depends on access pattern. 

So the criteria must be influenced both by user, which has its own business needs, and compaction strategy being used.

For TWCS, write amplification happens mostly on latest window, so it should always happen on fast tier. As access pattern differs for each business, it's important to allow user to specify the threshold for moving data into cold tier as proposed by Vlad.

For LCS, write amplification happens mostly on lower levels, so it's very important that they'll be on fast tier, to prevent slow tier from having its bandwidth saturated and compaction falling behind.
LCS and STCS mixes old and new data, so perhaps we'll have to allow their compaction to perform hot / cold data splitting into the respective directories.

The concept of tiered storage is very easy to implement for TWCS, so it's a very good candidate to start with. IIRC, Glauber visualized tiered storage with TWCS first, too.

Last but not least, I think tiered storage should be enabled on a per table basis, so users can allow a subset of tables to run on fast storage only if they see fit, whereas tables that are good candidates for tiered storage can have the feature enabled.

**Comment by rnz** [comment]
I think in same context - online moving data (keyspace, table) to other location.
for example:
https://kb.altinity.com/altinity-kb-setup-and-maintenance/altinity-kb-moving-table-to-another-device./
https://oracle-base.com/articles/12c/online-move-table-12cr2

**Comment by vladzcloudius** [comment]
> I think in same context - online moving data (keyspace, table) to other location. for example: https://kb.altinity.com/altinity-kb-setup-and-maintenance/altinity-kb-moving-table-to-another-device./ https://oracle-base.com/articles/12c/online-move-table-12cr2

This is an interesting feature but IMO orthogonal to this GH issue. 
You should probably file a separate GH issue for it with a description of a use case that would benefit from it.

**Comment by vladzcloudius** [comment]
> > > I think in same context - online moving data (keyspace, table) to other location. for example: https://kb.altinity.com/altinity-kb-setup-and-maintenance/altinity-kb-moving-table-to-another-device./ https://oracle-base.com/articles/12c/online-move-table-12cr2
> > 
> > 
> > This is an interesting feature, @rnz, but IMO orthogonal to this GH issue. You should probably file a separate GH issue for it with a description of a use case that would benefit from it.
> 
> I don't think so. Support for tiered storage in TWCS (and other compaction strategies) cannot exist without a basic data movement function, which would properly be implemented as a command available to the administrator.

I agree! This is exactly why I think it deserves a separate GH issue ;)
This API you refer is more generic than this GH issue because it (this GH issue) relates to a very specific option of implementing a tiered storage for a TWCS.
